### PR TITLE
[feature] support kvcache with static graph

### PR DESCRIPTION
**Pull request by xiaonans**
该提交使用 python llama_kvcache_inference.py 可验证 infinitensor 和 pytorch 在使用 kvcache 的时候精度是一致的。

### CODE REVIEW

**Comment by kilinchange** [comment]
结合晓楠给我的引入 attentionKVCache 算子之前和引入之后的 onnx 图，有几个问题想请教一下：
1. max_seq_len 是由 onnx 图给定吗？
2. Onnx 图里没有 attention mask？
3. input_1 是当前输入 token 的 id ？（之前跑的 llama 图这里是直接输入一个 [1,1,4096] 的向量，二者是等价的吗？）
4. 为什么要写成两个 kernel，各自分别承担什么功能呢？

File: src/kernels/cuda/attention_kvcache.cu [hunk 1]
```diff
@@ -2,127 +2,168 @@
...
-        ptr_sum[0] = 0;
-        #pragma unroll
-        for (int idx_SEQ_UNIT = 0; idx_SEQ_UNIT < SEQ_UNIT && idx_SEQ_UNIT + idx_seq < SEQ_LENGTH; idx_SEQ_UNIT ++) { 
-            ptr_P[idx_SEQ_UNIT] = expf(ptr_P[idx_SEQ_UNIT] - ptr_max[0]);
-            ptr_sum[0] += ptr_P[idx_SEQ_UNIT];
+    // softmax
+    #pragma unroll
+    for (int idx_SEQ_UNIT = 0; idx_SEQ_UNIT < SEQ_UNIT && idx_SEQ_UNIT + idx_seq < seq_length; idx_SEQ_UNIT ++) { 
+        ptr_P[idx_SEQ_UNIT] = expf(ptr_P[idx_SEQ_UNIT]);
+        ptr_sum[0] += ptr_P[idx_SEQ_UNIT];
```
Comment (xgqdut2016) [inline src/kernels/cuda/attention_kvcache.cu:79]: 这里计算softmax之前，不需要对ptr_P数组计算一次最大值，防止数值溢出吗？

**Comment by xiaonans** [comment]
> 结合晓楠给我的引入 attentionKVCache 算子之前和引入之后的 onnx 图，有几个问题想请教一下：
> 
> 1. max_seq_len 是由 onnx 图给定吗？
> 2. Onnx 图里没有 attention mask？
> 3. input_1 是当前输入 token 的 id ？（之前跑的 llama 图这里是直接输入一个 [1,1,4096] 的向量，二者是等价的吗？）
> 4. 为什么要写成两个 kernel，各自分别承担什么功能呢？

1. max_seq_len 就是 onnx 图当中 kvcache 那个 buffer 的大小。
2. 对，现在这个attention算子还不支持需要mask的情况，因为现在都是假设 Q 的 seq_len = 1，这种情况不需要 mask；未来这个算子还需要支持 mask。
3. 是的；之前跑的图我理解是泽众的图，我不确定泽众那边的输入是什么。
4. 为了加速计算，两个kernel接力跑，完成一个共同的任务，大的reduce的算子应该都是这样做的。

File: src/kernels/cuda/attention_kvcache.cu (hunk 1 above)
Comment (xiaonans) [reply src/kernels/cuda/attention_kvcache.cu:79]: > 这里计算softmax之前，不需要对ptr_P数组计算一次最大值，防止数值溢出吗？

其实是应该有的，但是现在还没遇上溢出的情况；后面还得找时间加上。你如果愿意的话，可以帮忙加上试试看精度和性能；直接拿我提交的脚本可以测一下端到端的精度，然后这个算子本身也有一个简单的测试可以粗略测一下精度。

**Review by kilinchange — APPROVED** [review]

**Review by whjthu — APPROVED** [review]
